pkg/storagepolicy/sdkstoragepolicy.go

```diff
@@ -87,82 +97,78 @@ func (p *SdkPolicyManager) Create(
-	// we need to marshal it into string using protobuf jsonpb
+	// we need to marshal sdkStoragePolicy into string using protobuf jsonpb
 	m := jsonpb.Marshaler{OrigName: true}
-	policyStr, err := m.MarshalToString(req.StoragePolicy.GetPolicy())
+	policyStr, err := m.MarshalToString(req.GetStoragePolicy())
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Json Marshal failed for policy %s: %v", req.StoragePolicy.GetName(), err)
+		return nil, status.Errorf(codes.Internal, "Json Marshal failed for policy %s: %v", req.GetStoragePolicy().GetName(), err)
 	}
 
-	_, err = p.kv.Create(prefixWithName(req.StoragePolicy.GetName()), policyStr, 0)
+	_, err = p.kv.Create(prefixWithName(req.GetStoragePolicy().GetName()), policyStr, 0)
 	if err == kvdb.ErrExist {
-		return nil, status.Errorf(codes.AlreadyExists, "Storage Policy already exist : %v", req.StoragePolicy.GetName())
+		return nil, status.Errorf(codes.AlreadyExists, "Storage Policy already exist : %v", req.GetStoragePolicy().GetName())
 	} else if err != nil {
 		return nil, status.Errorf(codes.Internal, "Failed to save storage policy: %v", err)
 	}
 
+	logrus.Infof("Storage Policy %v is created", req.GetStoragePolicy().GetName())
 	return &api.SdkOpenStoragePolicyCreateResponse{}, nil
 }
 
 // Update Storage policy
+// TODO: Decide whether update storage policy request should update ownership at all
+// We can check if user is admin then only allow update ownership
 func (p *SdkPolicyManager) Update(
 	ctx context.Context,
 	req *api.SdkOpenStoragePolicyUpdateRequest,
 ) (*api.SdkOpenStoragePolicyUpdateResponse, error) {
-	if req.StoragePolicy.GetName() == "" {
+	if req.GetStoragePolicy().GetName() == "" {
 		return nil, status.Error(codes.InvalidArgument, "Must supply a Storage Policy Name")
 	}
 
-	if req.StoragePolicy.GetPolicy() == nil {
+	if req.GetStoragePolicy().GetPolicy() == nil {
 		return nil, status.Error(codes.InvalidArgument, "Must supply Volume Specs")
 	}
 
-	// Get Existing details to merge
-	pol, err := p.Inspect(ctx,
+	// Get Existing details to check access
+	oldPolicy, err := p.Inspect(ctx,
 		&api.SdkOpenStoragePolicyInspectRequest{
-			Name: req.StoragePolicy.GetName(),
+			Name: req.GetStoragePolicy().GetName(),
 		},
 	)
 	if err != nil {
-		return nil, status.Errorf(codes.NotFound, "Policy with name %s not found: %v", req.StoragePolicy.GetName(), err)
+		return nil, err
 	}
 
-	// merge old policy spec with new
-	volSpecs := &api.VolumeSpecPolicy{}
 	m := jsonpb.Marshaler{OrigName: true}
-	// marshal old policy to volSpecs
-	oldSpec, err := m.MarshalToString(pol.StoragePolicy.GetPolicy())
+	updateStr, err := m.MarshalToString(req.GetStoragePolicy())
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Json Marshal failed for policy %s: %v", req.StoragePolicy.GetName(), err)
+		return nil, status.Errorf(codes.Internal, "Json Marshal failed for policy %s: %v", req.GetStoragePolicy().GetName(), err)
 	}
-	jsonpb.UnmarshalString(oldSpec, volSpecs)
 
```

Comment:
There is a missing check here. This call takes whatever ownership information they have updated and saves it as the new value, but this should not be. Instead Ownership should be checked if it was passed in and if it was, it needs to go through ownership.Update() so that it checks for the correct update rules, like, the system admin is the only one that can change the owner. If we allowed this call to go through as it is, for example, a collaborator with write access could send a new ownership object and become the owner.

Reply:
good catch, thanks will update